services/base.js

```diff
@@ -84,40 +88,71 @@ module.exports = class BaseService {
     return result;
   }
 
+  static _makeBadgeData(overrides, serviceData) {
+    const {
+      style,
+      label: overrideLabel,
+      logo: overrideLogo,
+      logoWidth: overrideLogoWidth,
+      link: overrideLink,
+      colorA: overrideColorA,
+      colorB: overrideColorB,
+    } = overrides;
+
+    const {
+      label: serviceLabel,
+      message: serviceMessage,
+      color: serviceColor,
+      link: serviceLink,
+    } = serviceData;
+
+    const defaultLabel = this.category;
+    const {
+      color: defaultColor,
+      logo: defaultLogo,
+    } = this.defaultBadgeData;
+
+    const badgeData = {
+      text: [
+        overrideLabel || serviceLabel || defaultLabel,
+        serviceMessage || 'n/a',
+      ],
+      template: isValidStyle(style) ? style : 'default',
+      logo: makeLogo(style === 'social' ? defaultLogo : undefined, { logo: overrideLogo }),
+      logoWidth: +overrideLogoWidth,
+      links: toArray(overrideLink || serviceLink),
+      colorA: makeColor(overrideColorA),
+    };
+    const color = makeColor(overrideColorB || serviceColor || defaultColor || 'lightgrey');
+    setBadgeColor(badgeData, color);
```

Comment:
This duplicates some of the code in `badge-data.js` but ultimately will replace it. I don't think it makes sense to reuse that function here; it just makes the code harder to reason about and more roundabout.

To make this even cleaner, we could rewrite some of these helper functions and/or change the interface to `makeBadge`.